[Codespaces] Disallow some operations on codespaces that have a pending operation

cwndrws opened the pull request:
Closes https://github.com/github/codespaces/issues/6794

We (the Codespaces team) are in the process of introducing the concept of pending async operations to codespaces. When a codespace has a pending operation, it can't be started, stopped, updated or exported.

This PR handles this scenario for each operation that is not allowed during a pending operation. It will show an error message to the user that the requested action isn't allowed because the codespace has a pending operation and shows the user the pending operation reason. See the individual commit message for each operation to see the exact changes to error messages, but one example is: 

Old ssh error message:

```
$ gh cs ssh -c cwndrws-redacted
Starting codespace ⣽error connecting to codespace: error starting codespace: HTTP 422: your codespace has an operation pending: updating to a sku with a different amount of storage; please wait until this operation is complete (https://api.github.com/user/codespaces/cwndrws-redacted/start)
exit status 1

```

New ssh error message:

```
$ gh cs ssh -c cwndrws-redacted
codespace is disabled while it has a pending operation: Changing machine types...
exit status 1
```

It also changes the way codespaces look in the list (`gh cs list`) to look disabled (gray) with the pending operation reason replacing its state.

![Screen Shot 2022-03-15 at 5 36 11 PM](https://user-images.githubusercontent.com/1382553/158477324-4c46c9a7-4a09-494f-9953-31c7bf1f9454.png)

cwndrws committed:
Add pending opertion and reason to codespace response struct

cwndrws committed:
Mark codespace with pending op as disabled with reason instead of state

cwndrws committed:
Disallow ssh'ing to codespace with a pending operation

Since the API already disallows this, this makes the error cleaner and
more explicit when a user is trying to start/ssh into a codespace that
has a pending operation.

Example of the old error message:

```
$ gh cs ssh -c cwndrws-redacted
Starting codespace ⣽error connecting to codespace: error starting codespace: HTTP 422: your codespace has an operation pending: updating to a sku with a different amount of storage; please wait until this operation is complete (https://api.github.com/user/codespaces/cwndrws-redacted/start)
exit status 1

```

Example of the new error message:

```
$ gh cs ssh -c cwndrws-redacted
codespace is disabled while it has a pending operation: Changing machine types...
exit status 1
```

cwndrws committed:
Disallow getting logs from codespaces with pending ops

Since the API already disallows this, this pretty much just cleans up
the error reporting to the user.

Example of old error:

```
$ gh cs logs -c cwndrws-redacted
Starting codespace ⣽connecting to codespace: error starting codespace: HTTP 422: your codespace has an operation pending: updating to a sku with a different amount of storage; please wait until this operation is complete (https://api.github.com/user/codespaces/cwndrws-redacted/start)
exit status 1

```

Example of new error:

```
$ gh cs logs -c cwndrws-redacted
codespace is disabled while it has a pending operation: Changing machine types...
exit status 1
```

cwndrws committed:
Ensure codespace exists and doesn't have a pending op when opening Code

The initial intention for this change was to disallow users to open a
codespace in VS Code if the codespace has a pending operation. This also
adds a side-benefit of presenting the user an error before waiting for
VS Code to open if they provide an invalid codespace to open.

cwndrws committed:
Disallow any port operations when codespace has pending operation

Since all of the port operations require the codespace to be running, we
need to disallow these operations when there's a pending op since we
can't start the codespace in this state.

Since the API already disallows this, this is basically cleaning up the
error messages that the user sees in this state

Old error message:

```
$ gh cs ports forward 80:80
? Choose codespace: redacted
Starting codespace ⣻error connecting to codespace: error starting codespace: HTTP 422: your codespace has an operation pending: updating to a sku with a different amount of storage; please wait until this operation is complete (https://api.github.com/user/codespaces/cwndrws-redacted/start)
```

New error message:

```
$ gh cs ports forward 80:80
? Choose codespace: redacted
codespace is disabled while it has a pending operation: Changing machine types...
exit status 1
```

cwndrws committed:
Add test for disallowing ssh when codespace has a pending op

cwndrws committed:
Add test for disallowing logs when codespace has a pending op

cwndrws committed:
Add test for disallowing opening vs code for codespace with pending op

I also refactored the existing vs code test a bit to work with the new
use of `getOrChooseCodespace`.

cwndrws committed:
Add tests for disallowing all port commands for codespace w/ pending op

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

josebalius commented on pkg/cmd/codespace/list.go:
```diff
@@ -89,10 +89,22 @@ func (a *App) List(ctx context.Context, limit int, exporter cmdutil.Exporter) er
 		tp.AddField(c.branchWithGitStatus(), nil, cs.Cyan)
-		tp.AddField(c.State, nil, stateColor)
+		if c.PendingOperation {
+			tp.AddField(c.PendingOperationDisabledReason, nil, cs.Gray)
```
nit: use `nameColor` to keep them consistent in case someone changes the Gray to something else?

cwndrws replied:
Great Idea!

josebalius commented on pkg/cmd/codespace/ports.go:
```diff
@@ -380,3 +370,23 @@ func normalizeJSON(j []byte) []byte {
 	return bytes.ReplaceAll(j, []byte("},}"), []byte("}}"))
 }
+
+func getCodespaceForPorts(ctx context.Context, apiClient apiClient, codespaceName string) (*api.Codespace, error) {
```
Wondering if we shouldn't name this something more generic and use it everywhere else to DRY things up a bit?

cwndrws replied:
Yeah I think that's a good idea or just move the error handling into `getOrChooseCodespace` since I think we're now doing this error handling at every `getOrChooseCodespace` call site.

josebalius replied:
Yeah that works!

josebalius reviewed (approved):
✨ love the tests. Left a couple of comments nothing blocking.

cwndrws committed:
Use color variable instead of literal for disabled reason

cwndrws committed:
Refactor the getOrChooseCodespace to always check for pending ops

adonovan commented on pkg/cmd/codespace/code.go:
```diff
@@ -31,18 +31,12 @@ func newCodeCmd(app *App) *cobra.Command {
-		codespaceName = codespace.Name
+	codespace, err := getOrChooseCodespace(ctx, a.apiClient, codespaceName)
+	if err != nil {
+		return fmt.Errorf("get or choose codespace: %w", err)
```
This isn't a particularly user-friendly error message. "get or choose codespace" is really just a function name, rendered with spaces: it's an internal detail of the structure of the program. The error message from getOrChooseCodespaces should be good enough to explain itself, so that you can simply return err here (as you do later on in this PR).

cwndrws replied:
I agree. I just copied this from the other places where we call `getOrChooseCodespace`. I'll go through and change all of them.

adonovan commented on internal/codespaces/api/api.go:
```diff
@@ -168,17 +168,19 @@ func (a *API) GetRepository(ctx context.Context, nwo string) (*Repository, error
 
 // Codespace represents a codespace.
```
Can you add a link to the place where the meaning of all of these fields is specified?

cwndrws replied:
You mean add a link to the API docs? The fields I added in this PR aren't publicly documented yet since they are still behind a feature flag. I can add a link to the public API docs, but it won't have the two new fields in it.

adonovan replied:
Yep, the API doc. It's fine if it's not entirely visible, or even visible only to Hubbers, but it should be possible for maintainers to "pull on the thread".

cwndrws committed:
Fix copy pasta error to appease the linter

cwndrws committed:
Remove unhelpful error wrapper

cwndrws committed:
Disallow editing a codespace that has a pending operation

This is already prevented by the API, but this will make the error
message cleaner and easier to understand for the user.

Example of old error message:

```
$ gh cs edit -c cwndrws-redacted -d "some silly name"
error editing codespace: HTTP 422: your codespace has an operation pending: updating to a sku with a different amount of storage; please wait until this operation is complete (https://api.github.com/user/codespaces/cwndrws-redacted)
exit status 1

```

Example of new error message:

```
$ gh cs edit -c cwndrws-redacted -d "some silly name"
error editing codespace: codespace is disabled while it has a pending operation: Changing machine types...
exit status 1
```

cwndrws committed:
Add docs link comment to Codespaces struct definition
